File: androidApp/src/main/kotlin/br/com/qualorock/androidApp/ui/viewmodel/EmailVerificationViewModel.kt
package br.com.qualorock.androidApp.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import data.QorConfig
import domain.user.VerifyEmailResult
import domain.user.usecase.VerifyEmail
import kotlinx.coroutines.Job
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

private const val MillisPerSecond = 1_000L
private const val OtpCodeLength = 6

/** Client-side validation for the 6-digit OTP field — non-empty + exact length only (AUTH-10). */
enum class CodeFieldError {
    Required,
    InvalidLength,
}

/** One-shot outcome [EmailVerificationScreen] consumes exactly once. Nav-graph wiring happens in A14. */
sealed class EmailVerificationEvent {
    data object Verified : EmailVerificationEvent()
}

/**
 * Pure resend-cooldown state machine — [br.com.qualorock.androidApp.ui.components.CtaButtonMotion]-style
 * pure function so ticking is unit-testable without Compose or a real `delay()`.
 * [EmailVerificationViewModel] drives it with a `viewModelScope` coroutine ticker; [tick] and
 * [started] carry no side effects of their own.
 */
data class VerificationCooldown(val remainingSeconds: Long) {
    /** True once the countdown has fully elapsed — the resend action becomes tappable. */
    val canResend: Boolean get() = remainingSeconds <= 0

    fun tick(): VerificationCooldown = VerificationCooldown((remainingSeconds - 1).coerceAtLeast(0))

    companion object {
        fun started(totalSeconds: Long = QorConfig.EmailVerificationResendCooldownSeconds): VerificationCooldown =
            VerificationCooldown(totalSeconds)
    }
}

/** Form + submission state for [EmailVerificationScreen]. */
data class EmailVerificationUiState(
    val code: String = "",
    val codeError: CodeFieldError? = null,
    val submitError: String? = null,
    val isLoading: Boolean = false,
    val cooldown: VerificationCooldown = VerificationCooldown.started(),
    val resendConfirmation: Boolean = false,
)

/**
 * A9 — form state + submit orchestration for `EmailVerificationScreen` (AUTH-01/AUTH-10).
 * Delegates to [VerifyEmail] (S12b's thin wrapper over `UserRepository`); owns client-side OTP
 * validation, loading/error UI state, the [VerificationCooldown] ticker driving the resend
 * button, and the one-shot [events] the caller's nav graph wires up in A14.
 *
 * The `email` this screen verifies is nav-graph state (A14), not this class's own — [onSubmit]
 * and [onResend] take it as a parameter, matching the DI registration `EmailVerificationViewModel(get())`
 * in `AppModule.kt` (only [VerifyEmail] is Koin-injected).
 *
 * **Verifying does NOT log the fan in.** Per [VerifyEmailResult]'s KDoc, `verifyEmailCode` only
 * marks the account verified — it returns no session/token — so [EmailVerificationEvent.Verified]
 * must be wired (A14) to navigate to Login, never straight to Home; this matches `qor-website`'s
 * shipped W20 behavior (see `verificar-email/page.tsx`, which links to `/entrar` after success).
 *
 * [VerifyEmail.resend] is fire-and-forget (`Unit`, no server message) — a resend tap always shows
 * a static local pt-BR confirmation string and restarts the cooldown, it never expects text back
 * from the use case.
 */
class EmailVerificationViewModel(private val verifyEmail: VerifyEmail) : ViewModel() {

    private val _uiState = MutableStateFlow(EmailVerificationUiState())
    val uiState: StateFlow<EmailVerificationUiState> = _uiState.asStateFlow()

    private val _events = Channel<EmailVerificationEvent>(Channel.BUFFERED)
    val events: Flow<EmailVerificationEvent> = _events.receiveAsFlow()

    private var cooldownJob: Job? = null

    init {
        startCooldown()
    }

    fun onCodeChange(value: String) {
        _uiState.update { it.copy(code = value, codeError = null, submitError = null) }
    }

    fun onSubmit(email: String) {
        val state = _uiState.value
        val codeError = validateCode(state.code)
        if (codeError != null) {
            _uiState.update { it.copy(codeError = codeError) }
            return
        }

        _uiState.update { it.copy(isLoading = true, submitError = null) }
        viewModelScope.launch {
            when (val result = verifyEmail.verifyCode(email, state.code)) {
                is VerifyEmailResult.Success -> {
                    _uiState.update { it.copy(isLoading = false) }
                    _events.send(EmailVerificationEvent.Verified)
                }

                is VerifyEmailResult.Failure -> {
                    _uiState.update { it.copy(isLoading = false, submitError = result.message) }
                }
            }
        }
    }

    fun onResend(email: String) {
        if (!_uiState.value.cooldown.canResend) return
        viewModelScope.launch {
            verifyEmail.resend(email)
            _uiState.update { it.copy(resendConfirmation = true) }
            startCooldown()
        }
    }

    private fun startCooldown() {
        cooldownJob?.cancel()
        _uiState.update { it.copy(cooldown = VerificationCooldown.started()) }
        cooldownJob = viewModelScope.launch {
            while (_uiState.value.cooldown.remainingSeconds > 0) {
                delay(MillisPerSecond)
                _uiState.update { it.copy(cooldown = it.cooldown.tick()) }
            }
        }
    }

    private fun validateCode(code: String): CodeFieldError? = when {
        code.isBlank() -> CodeFieldError.Required
        code.length != OtpCodeLength -> CodeFieldError.InvalidLength
        else -> null
    }

    override fun onCleared() {
        cooldownJob?.cancel()
    }
}

File: iosApp/iosApp/UI/Screens/EmailVerificationViewModel.swift
import Foundation
import shared

/// I9 — client-side validation for the 6-digit OTP field, non-empty + exact length only
/// (AUTH-10). Mirrors Android's `CodeFieldError`.
enum CodeFieldError: Equatable {
    case required
    case invalidLength

    var message: String {
        switch self {
        case .required:
            return String(localized: "email_verification_error_code_required")
        case .invalidLength:
            return String(localized: "email_verification_error_code_invalid_length")
        }
    }
}

/// Which flow launched `EmailVerificationView` — the real endpoint issues no session on
/// verify (see [EmailVerificationViewModel]'s doc), so where `onVerified` should navigate
/// differs by caller: back to Login after signup, back to Profile after an email-change.
/// Nav-graph state, like `email` — not owned by this screen's own state.
enum EmailVerificationContext {
    case signup
    case emailChange
}

/// Pure resend-cooldown state machine, mirroring Android's `VerificationCooldown` — a plain
/// value type so ticking is unit-testable without any async waiting.
struct VerificationCooldown: Equatable {
    var remainingSeconds: Int

    /// True once the countdown has fully elapsed — the resend action becomes tappable.
    var canResend: Bool { remainingSeconds <= 0 }

    func ticked() -> VerificationCooldown {
        VerificationCooldown(remainingSeconds: max(remainingSeconds - 1, 0))
    }

    static func started(
        totalSeconds: Int = Int(QorConfig.shared.EmailVerificationResendCooldownSeconds)
    ) -> VerificationCooldown {
        VerificationCooldown(remainingSeconds: totalSeconds)
    }
}

/// Form + submission state for `EmailVerificationView`.
struct EmailVerificationUiState: Equatable {
    var code: String = ""
    var codeError: CodeFieldError?
    var submitError: String?
    var isLoading: Bool = false
    var cooldown: VerificationCooldown = .started()
    var resendConfirmation: Bool = false
}

/// Narrow protocol over `shared`'s `VerifyEmail` use case so tests can substitute a fake
/// without touching Koin — `VerifyEmail` itself is `final` on the Kotlin/Native side (can't be
/// subclassed from Swift), but its designated initializer takes the `UserRepository` protocol,
/// so a real `VerifyEmail(userRepository: fake)` built from a fake repository conforms to this
/// exactly the same way Android's `EmailVerificationViewModelTest` builds a real `VerifyEmail`
/// over a `FakeVerifyEmailUserRepository`.
protocol EmailVerifying {
    func resend(email: String) async throws
    func verifyCode(email: String, code: String) async throws -> VerifyEmailResult
}

extension VerifyEmail: EmailVerifying {}

private let otpCodeLength = 6

/**
 * I9 — form state + submit orchestration for `EmailVerificationView` (AUTH-01/AUTH-10).
 * Delegates to `VerifyEmail` (S9's thin wrapper over `UserRepository`); owns client-side OTP
 * validation, loading/error UI state, and the `VerificationCooldown` driving the resend button.
 *
 * **Verifying does NOT log the fan in.** Per `VerifyEmailResult`'s KDoc, `verifyEmailCode` only
 * marks the account verified — it returns no session/token — so this must never navigate
 * straight to Home. Which destination it *does* navigate to depends on which flow launched this
 * screen (`context`, passed in like `email` as nav-graph state, not this class's own): signup
 * routes to Login (matching `qor-website`'s shipped W20 behavior), an email-change routes back
 * to Profile. [onVerifiedForSignup]/[onVerifiedForEmailChange] are therefore two distinct
 * closures rather than one generic "success" callback, so I14's nav-graph wiring can tell them
 * apart — this view never pushes navigation itself.
 *
 * `resend` is fire-and-forget (no server message) — a resend tap always shows a static local
 * pt-BR confirmation string and restarts the cooldown; a failure is swallowed rather than shown,
 * matching Android's `EmailVerificationViewModel.onResend`.
 */
@MainActor
final class EmailVerificationViewModel: ObservableObject {
    @Published private(set) var uiState = EmailVerificationUiState()

    private let context: EmailVerificationContext
    private let verifyEmail: EmailVerifying
    private let onVerifiedForSignup: () -> Void
    private let onVerifiedForEmailChange: () -> Void
    private var cooldownTask: Task<Void, Never>?

    init(
        context: EmailVerificationContext,
        verifyEmail: EmailVerifying = IosDependencies.shared.verifyEmail(),
        onVerifiedForSignup: @escaping () -> Void,
        onVerifiedForEmailChange: @escaping () -> Void
    ) {
        self.context = context
        self.verifyEmail = verifyEmail
        self.onVerifiedForSignup = onVerifiedForSignup
        self.onVerifiedForEmailChange = onVerifiedForEmailChange
        startCooldown()
    }

    deinit {
        cooldownTask?.cancel()
    }

    func onCodeChange(_ value: String) {
        uiState.code = value
        uiState.codeError = nil
        uiState.submitError = nil
    }

    func onSubmit(email: String) async {
        if let codeError = Self.validate(code: uiState.code) {
            uiState.codeError = codeError
            return
        }

        uiState.isLoading = true
        uiState.submitError = nil
        let code = uiState.code

        do {
            let result = try await verifyEmail.verifyCode(email: email, code: code)
            uiState.isLoading = false
            if result is VerifyEmailResult.Success {
                switch context {
                case .signup:
                    onVerifiedForSignup()
                case .emailChange:
                    onVerifiedForEmailChange()
                }
            } else if let failure = result as? VerifyEmailResult.Failure {
                uiState.submitError = failure.message
            }
        } catch {
            uiState.isLoading = false
            uiState.submitError = error.localizedDescription
        }
    }

    func onResend(email: String) async {
        guard uiState.cooldown.canResend else { return }
        try? await verifyEmail.resend(email: email)
        uiState.resendConfirmation = true
        startCooldown()
    }

    private func startCooldown() {
        cooldownTask?.cancel()
        uiState.cooldown = .started()
        cooldownTask = Task { [weak self] in
            while let self, self.uiState.cooldown.remainingSeconds > 0 {
                try? await Task.sleep(nanoseconds: 1_000_000_000)
                if Task.isCancelled { return }
                self.uiState.cooldown = self.uiState.cooldown.ticked()
            }
        }
    }

    /// Test-only escape hatch so `onResend`'s cooldown-gated path is unit-testable without a
    /// real `Task.sleep` wait — mirrors Android's `EmailVerificationViewModelTest` using a fake
    /// clock/dispatcher instead.
    func forceCooldownElapsedForTesting() {
        cooldownTask?.cancel()
        uiState.cooldown = VerificationCooldown(remainingSeconds: 0)
    }

    private static func validate(code: String) -> CodeFieldError? {
        if code.isEmpty { return .required }
        if code.count != otpCodeLength { return .invalidLength }
        return nil
    }
}
